### ISSUE TITLE
Successful creation of DNN models for Julius and Polish

### ISSUE THREAD

**Issue by palles77** [julius-speech/julius#68]
Hi All

Just to let anyone interested to know, that after 8 months of in depth research I managed to convert HTK 3.5.2 models to Julius DNN format and received really good results. Example WER reduction was f2.5% from 17.5% to 15% on 112h of audio for Polish LVCSR task. I used 48 MFCC vector and 1536 nodes 5 layer DNN model. Now will be doing experiments for bigger networks.

Leslaw

**Comment by araczkowski** [comment]
Hi Lesław,

interesting... is the any way to see/test your results?
Do you have it somewhere in github?
For my project I'm looking for the good offline speech to text solution in Polish.

Andrzej

**Comment by palles77** [comment]
Current version is available on sf.net in skrybotdomowy.sf.net. There are raw Julius speech models there available too. I will soon make a new version available and you will be able to test DNNs using C# in your own projects. Obviously here I am talking about open source speech models. If you are interested in commercial quality models you can contact me via skrybot.pl

**Comment by r1me** [comment]
Will you create some day a model for CMU Sphinx by any chance ?

**Comment by palles77** [comment]
No I don't see that happening. It took me 12 years to learn how to get HTK and Julius work perfectly, so not really planning on taking on Sphinx, even though I considered it in the past. My entire loyalty lies with Julius as for me it is the best decoder out there.

**Comment by NaveenDeshpande** [comment]
Hi Lechu,


Greetings !!


I am using Julius speech recognition engine for voice recognition in my project, i have installed it and everything is fine it is able to recognize the voice commands given as input through microphone but the problem i am facing is, i am not able to control it completely. i have few doubts regarding Julius :


1) I want to know is there any method to make Julius recognize voice only once instead of recognizing continuously. (one command it should recognize and stop).

2) Even if i don't speak anything through microphone, for little sound also it recognizes and gives something present in the dictionary file, why is it like that ? what can i do to control it?

3) How to control it ? for ex. if i say play song it should only take play song as input not any other commands, when i say play sometimes it will recognize different commands like pause or mute etc.

4) Is there any way to avoid noise and make it only take the necessary commands given as input through microphone?


Please can you guide me with the possible solution ASAP.


Thanks & Regards

Gururaj Deshpande


________________________________

**Comment by palles77** [comment]
This is a wrong subtopic to ask this question. You should open a new issue for that. However to quickly answer: Julius is just a decoder, it was not meant to be a suite which supports all the things you mentioned. Incidentally Julius does have code which deals with short pause segmentation, silence detection and communication via threads. You should really download the codebase, run the examples and read JuliusBook which is available https://osdn.net/projects/julius/downloads/47534/Juliusbook-4.1.5.pdf/

**Comment by NaveenDeshpande** [comment]
Thanks for your reply.

I don't know much about github i am new to it can you please guide me in this matter.

There are few problems i am facing, i didn't get replies in voxforge forum so i am asking here.


1)I am using live microphone to give the inputs to julius, the problem i am facing is...whenever i give input through the live microphone, for example: the command is play song and if this is not present in the .dict file, even though julius will return some random command which is present in the .dict file...i want to know how to control this and  make  it more flexible..like i want julius to return nothing or search failed command as output if the given input is not present in the .dict file.

2) For small sound also julius recognizes and gives the result even if i have not said a single word through the microphone. how to avoid this?

________________________________

**Comment by palles77** [comment]
1) It depends on quality of your speech model and recording. It is a huge topic. Looks like you should be using grammars and filter out the result based on confidence of the recognized word. Julius supports print out of N best hypotheses, so you could use that. Try looking at issuing commands to Julius via its socket interface and look at the results it provides. The results are very detailed and it is likely you will find some help there.
2) This is speech/non speech detection which Julius has support for, but again very difficult topic. Try to search for projects based on Julius. I know there are some Raspberry Pi versions of Julius and for home automation. They may solve your problem. I personally do not use those and write my own extensions to Julius.

**Comment by NaveenDeshpande** [comment]
you said :

Julius supports print out of N best hypotheses, so you could use that.


how can i do this ? i tried using the confidence level but didn't get the required output  for that. what else i can try to make it more flexible.


can i find projects on julius in github ?


Thanks


________________________________

**Comment by palles77** [comment]
You should use Visual Studio to debug Julius. That is what I do. Otherwise you can add your own jlog() calls to Julius's code. You can download VS for free and Julius supports Windows version as well.

**Comment by G10DRAS** [comment]
@palles77 how did you generate lexicon and language model for julius-dnn ?

**Comment by palles77** [comment]
@G10DRAS I used HTK book language model recipe. HTK language model tools can be used for generating forward 3-gram LM ARPA model and backwards 2-gram LM ARPA model. They can be both combined into one file using mkbingram. For lexicon I used a pronunciation generator written by myself in Perl.

**Comment by G10DRAS** [comment]
@palles77 I am able to generate language model using mkbingram and its working good, but stuck at lexicon part. 
Is your pronunciation generator script available with JuliusModels ?

**Comment by G10DRAS** [comment]
@palles77 I wrote the pronunciation generator script and now everything works perfect.
Thanks for JuliusModels.

**Comment by palles77** [comment]
@G10DRAS Yes - JuliusModels is quite a breakthrough. I am going to update it over time with even better models. Glad that your pronunciation generator works. Keep in mind that around 30 000 words in Polish models are have a manual pronunciation done for them (specifically foreign sounding words).

**Comment by welliX** [comment]
> Hi All
> 
> Just to let anyone interested to know, that after 8 months of in depth research I managed to convert HTK 3.5.2 models to Julius DNN format and received really good results. Example WER reduction was f2.5% from 17.5% to 15% on 112h of audio for Polish LVCSR task. I used 48 MFCC vector and 1536 nodes 5 layer DNN model. Now will be doing experiments for bigger networks.
> 
> Leslaw

Hi Leslaw,
is there any (step-by_step) recipe, tutorial, script, or similar available, that describes how the Julius DNN acoustic models are created (from/via HTK?)?
many thanks for any help!
   kind regards, Andi
